pygmt/clib/session.py

```diff
@@ -1098,8 +1169,28 @@ def virtualfile_from_vectors(self, *vectors):
             family, geometry, mode="GMT_CONTAINER_ONLY", dim=[columns, rows, 1, 0]
         )
 
+        # Use put_vector for columns with numerical type data
         for col, array in enumerate(arrays):
-            self.put_vector(dataset, column=col, vector=array)
+            if col not in str_cols:
+                self.put_vector(dataset, column=col, vector=array)
+
+        # Use put_strings for last column(s) with string type data
+        # Have to use modifier "GMT_IS_DUPLICATE" to duplicate the strings
+        if str_cols:
+            if len(str_cols) == 1:
+                strings = arrays[str_cols[0]]
+            elif len(str_cols) == 2:
+                strings = np.char.add(
+                    np.char.add(arrays[str_cols[0]], " "), arrays[str_cols[1]]
+                )
+            else:
+                raise NotImplementedError(
+                    f"Unable to handle {len(str_cols)} columns of string arrays. "
+                    "Please use only 1 or 2 instead."
+                )
+            self.put_strings(
+                dataset, family="GMT_IS_VECTOR|GMT_IS_DUPLICATE", strings=strings
+            )
```

Comment:
Ok, I've handled the issue at https://github.com/GenericMappingTools/pygmt/pull/520#discussion_r462591754 (was trickier than I thought) and also added some tests, Up to 2  'str' type columns can be passed into `virtualfile_from_vectors` now.

I've asked a question on stackoverflow at https://stackoverflow.com/questions/63332200/concatenate-multiple-numpy-string-arrays-with-a-space-delimiter to handle `n>2` cases, may refactor this later.

Reply:
Ok, refactored at 4f1ccda5a736ff807f68be1c6462dc7eda0a30cd.